woocommerce/woocommerce-ios
# [Bundle products in order form] Validate all bundle items and display error message

Part of #10428 


## Why


When saving the configuration for a bundle product in the order form, there are a few checks to validate before sending it remotely. Otherwise, the remotely updated order will result in just the bundle product order item without any child bundled items (the configuration fails silently). This PR validates all the bundle items. I plan to add unit tests later to focus on other more challenging subtasks first.

## How

The validation logic is documented in Pe5pgL-3Vd-p2#validation-of-bundle-configuration. When validating each bundle item, an error message is shown when there's a validation error.

🗒️ I also considered a different UX to disable the Done CTA when any of the bundle item configurations is invalid. However, after a few attempts, I couldn't get to observe the "isValid: Bool" publisher from nested bundle item view models as changes aren't emitted when the observable property of a nested observable object changes. Thus, I decided to go with the UX where the Done CTA is always enabled and validation checks are performed on tap.

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with non-empty items and **at least one variable item with a variation that has "Any" attribute and another variation that doesn't have "Any" attributes**.

- Go to the Orders tab
- Tap `+` to create an order
- Tap `Add Products` and select a bundle product with at least one variable item --> the bundle product should be added to the order
- Tap on the expand  🔽 CTA on the bundle product order item
- Tap `Configure` --> after a bit (products being loaded remotely), the configurable bundle items should be shown
- Tap `Done` without selecting a variation for the variable item --> an error message should be shown about the missing variation
- Tap `Select variation` on the variable item and select a variation that has an "Any" attribute
- Tap `Done` --> an error message should be shown about the missing variation attributes
- Tap `Select variation` on the variable item and select a variation that doesn't have "Any" attribute
- Tap `Done` --> the bundle configuration should be saved to the order item remotely

## Screenshots


missing variation | missing some attribute values
-- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/d4a8b776-3335-425b-8770-8c3c89777802" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/3bd4c6a6-84a9-4452-99ee-7eab321901c0" width="300" />


https://github.com/woocommerce/woocommerce-ios/assets/1945542/163795df-6bd8-4c08-9f07-328fd22a9488



---
